source/AI.cpp

```diff
@@ -2061,95 +2068,191 @@ void AI::KeepStation(Ship &ship, Command &command, const Body &target)
+			double outfitSpace = (weapon->Get("outfit space") + weapon->Get("weapon capacity")
+					+ weapon->Get("engine capacity")) / -2;
+			totalSpace += outfitSpace;
 
 			// Exploding weaponry that can damage this ship requires special
 			// consideration (while we have the ammo to use the weapon).
-			if(hasThisAmmo && weapon->BlastRadius() && !weapon->IsSafe())
-				minSafeDistance = max(weapon->BlastRadius() + weapon->TriggerRadius(), minSafeDistance);
+			if(hasThisAmmo && weapon->SafeRange())
+			{
+				minSafeDistance = max(weapon->SafeRange(), minSafeDistance);
+				splashSpace += outfitSpace;
+			}
 
-			// The missile boat AI should be applied at 1000 pixels range if
-			// all weapons are homing or turrets, and at 2000 if not.
-			double multiplier = (hardpoint.IsHoming() || hardpoint.IsTurret()) ? 1. : .5;
-			shortestRange = min(multiplier * weapon->Range(), shortestRange);
+			// The artillery AI should be applied at 1000 pixels range.
+			// regardless of wether the weapon is homing or not.
+			// The AI works fine with non homing weapons
+			double range = weapon->Range();
+			shortestRange = min(range, shortestRange);
+			if(range > 1000)
+			{
+				shortestArtillery = min(range, shortestArtillery);
+				rangedSpace += outfitSpace;
+			}
+			
+			
 		}
 	}
+	
+	// Calculate this ship's "turning radius"; that is, the smallest circle it
+	// can make while at full speed.
+	double stepsInFullTurn = 360. / ship.TurnRate();
+	double circumference = stepsInFullTurn * ship.Velocity().Length();
+	double diameter = circumference / PI;
+	
 	// If this ship was using the missile boat AI to run away and bombard its
 	// target from a distance, have it stop running once it is out of ammo. This
 	// is not realistic, but it's a whole lot less annoying for the player when
 	// they are trying to hunt down and kill the last missile boat in a fleet.
 	if(isArmed && !hasAmmo)
+	{
 		shortestRange = 0.;
+		shortestArtillery = 0.;
+	}
+	
+	// ArtilleryAI is the AI responsible for handling the behavior of missile boats
+	// and other ships with exceptionally long range weapons such as detainers
+	// The AI shouldn't use artilleryAI if it has no reverse and it's turning
+	// capabilities are very bad. Otherwise it spends most of it's time flying around
+	if(rangedSpace > totalSpace * 0.5 && (ship.MaxReverseVelocity() ||
+			diameter < 0.2 * shortestArtillery))
+		artilleryAI = true;
+	// Don't try to avoid your own splash damage if it means you whould be losing out
+	// on a lot of DPS. Helps with ships with very slow turning and not a lot of splash
+	// weapons being overly afraid of dying.
+	if(minSafeDistance && !(artilleryAI || shortestRange * splashSpace / totalSpace
+			> diameter))
+		minSafeDistance = 0.;
+
+	
+	ship.SetArtilleryAI(artilleryAI);
+	ship.SetShortestRange(shortestRange);
+	ship.SetShortestArtillery(shortestArtillery);
+	ship.SetMinSafeDistance(minSafeDistance);
+	ship.SetTurningRadius(diameter);
```

Comment:
These values don't belong on the Ship model, [they belong on a new class object that has a single responsibility](https://discord.com/channels/251118043411775489/285540320823869441/929552091368804362).

Add the new class in an "ai" subfolder whose job is to hold these things. You may find that it makes a lot of sense to want to move existing fields into that class too, but save that for a follow-up.

(Computing these values could certainly be done there, in a method whose signature is e.g. `void UpdateCache(const Ship &)`.)

Reply:
The newly added ShipAICache files should fix this!